dxp/demystifying_aui_script_and_liferay-frontend_component_tags.md

```diff
@@ -0,0 +1,114 @@
+
+```jsp
+<aui:script require="metal-dom/all/dom.js as dom">
+
+	function myFunction(element) {
+		dom.addClasses(element, 'hide');
+	}
+
+	Liferay.once('something', myFunction);
+
+</aui:script>
+```
+
+#### Settling the way which the script will be placed on the page using `position`:
+
+This property will change the way in which / how the script will be injected into the page. By default, those scripts that are wrapped by `aui:script` will be added to the minified files as described here (Pure JS on the JSP [PUT THE LINK HERE OH]).
+
+When placing the option of `inline` for this property, the scripts will be placed in a script tag obeying the order of the DOM where the JSP is being rendered. See [here](https://github.com/liferay/liferay-portal/blob/master/util-taglib/src/com/liferay/taglib/aui/ScriptTag.java#L143..L165).
+
+#### `sandbox`:
+
+The purpose for the sandbox attribute is to wrap the contents of the script tag in a self-invoking function, and to enable short-handing the `$` and `_` variables rather than calling `AUI.$` and `AUI._`.
+
+#### Caveats:
+
+We couldn’t use both `require` and `use` in the same file. Because `use` property calls the modularization for alloy-ui modules and `require` follows ES6 modularized files.
+
+You can see more information about `aui:script` [here](https://help.liferay.com/hc/en-us/articles/360017882752-Loading-Modules-with-AUI-Script-in-Liferay-DXP).
+
+## `liferay-frontend:component` tag
+
+We can use this tag for removing JavaScript code from JSPs. This tag accepts a `module` property wether we can pass a relative path for the script to be loaded. Also, this tag registers automatically the component to [Liferay’s component global registry](https://github.com/liferay/liferay-frontend-guidelines/blob/master/dxp/liferay_component.md#register), depending on the `componentId` property. When using this tag, `namespace`(or `portletNamespace`) and `spritemap` props will turn available.
+
+Here is an example of usage of this tag:
+
+On the JSP context:
+
+```jsp
+
+...
+<liferay-frontend:component
+	componentId="myComponent"
+	module="path/to/my/js/myComponent.js"
+/>
+...
+```
+
+And when exporting the `myComponent.js` file we can define a default function there like:
+
+```js
+export default function myComponent(props) {
+	console.log(props.namespace);
+	// it will return the portlet namespace where the liferay-frontend:component tag were called
+	console.log(props.spritemap);
+	// it will return the path to the page's spritemap
+}
+```
+
+This taglib can be very handy if you want to use modern ES6 features, since We don’t have support for transpiling ES6 from taglibs.
+
+Also, this tag adds some treatments for handling JavaScript on our JSPs like [adding special treatments for our built-in SPA framework](https://github.com/liferay/liferay-portal/blob/master/modules/apps/frontend-taglib/frontend-taglib/src/main/java/com/liferay/frontend/taglib/servlet/taglib/ComponentTag.java#L225).
```

Comment:
- "treatments" → "treatment" (in both places)
- "on" → "in"

If you are going to link to a specific line, you should use an "evergreen" link (ie. one with a hash instead of a symbolic name like `master`); you can get that by pressing "y" on GitHub when you're looking at the page in question.

Reply:
> If you are going to link to a specific line, you should use an "evergreen" link 

I asked myself about it when writing and I decided to adopt `master` as the branch to retrieve an updated doc. However, makes sense using "evergreen" for all the cases but it require more maintenance on docs for updating the branch and see if this info still valid.

I'll update it :p

Reply:
The problem with `master` links (or any link to any symbolic ref, like a branch name), is that they can break when files move. The worst class of links are those that link to a specific line, because lines can move around within a file. If a file moves, the user will get a 404 and think, "Oh, a busted link", but if lines move, the user will land on a page and think, "Huh? I'm not seeing anything like I expected to see", which is worse.

We discussed evergreen links in the past [here](https://github.com/liferay/liferay-frontend-guidelines/pull/201#discussion_r487035366), but I think it's time to document this, so I made:

https://github.com/liferay/liferay-frontend-guidelines/pull/228